### DISCUSSION TITLE
Writing to LanceDB in cloud object storage while other processes are reading?

### DISCUSSION THREAD

**Discussion by metadaddy** [lancedb/lancedb#1888] [Q&A] [answered]
Hi there - I'm writing a sample app to show how you can add docs to a LanceDB database in cloud object storage ([Backblaze B2](https://www.backblaze.com/cloud-storage) in my case) over time. I just want to confirm that my understanding of concurrent operations is correct...

Concurrent writes are not safe. I know this from the `WARN  lance_table::io::commit] Using unsafe commit handler. Concurrent writes may result in data loss. Consider providing a commit handler that prevents conflicting writes.` message the first time I write data. That's fine - I can ensure that only one process/thread is writing to the database at any given time.

My question is, how are readers affected when a write takes place? I assume the write happens in such a way that readers always see a consistent database - there is no way to read the database mid-write and receive inconsistent data.

Is this correct?

**Comment by wjones127** [comment, marked as answer]
> how are readers affected when a write takes place? I assume the write happens in such a way that readers always see a consistent database - there is no way to read the database mid-write and receive inconsistent data.

Readers will be unaffected. We use "Multi-version concurrency control". That means each write to the database creates a new version of the table. Readers will be pointing to older versions of the table, and thus won't be affected.

**Reply by metadaddy** [reply]
Many thanks for the confirmation!
